readthedocs/search/views.py

```diff
@@ -68,26 +80,9 @@ def elastic_search(request, project_slug=None):
             if value:
                 kwargs[avail_facet] = value
 
-        if user_input.type == 'project':
-            search = ProjectSearch(
-                query=user_input.query, user=request.user, **kwargs
-            )
-
-        elif user_input.type == 'domain':
-            search = DomainSearch(
-                query=user_input.query, user=request.user, **kwargs
-            )
-
-        elif user_input.type == 'file':
-            search = PageSearch(
-                query=user_input.query, user=request.user, **kwargs
-            )
-
-        elif user_input.type == 'all':
-            search = AllSearch(
-                query=user_input.query, user=request.user, **kwargs
-            )
-
+        search = search_facets[user_input.type](
+            query=user_input.query, user=request.user, **kwargs
```

Comment:
The solution with defaultdict is more readable to me, I like it. Although, there is a new logic hidden: if the type does not exist (for any reason) we are going to return `ProjectSearch` results. Is that OK? would be better to just return no results at all?

Reply:
We may want to add another condition next to the `user_input.query`

    ... and (user_input.type in search_facets):

and remove the lambda from the defaultdict.

Reply:
`project` is the default https://github.com/rtfd/readthedocs.org/blob/ccde7c598e2009e0fdb17505cd68c36153836376/readthedocs/search/views.py#L54

Reply:
:+1: 

It seems we have the default set in multiples places then. I think we could just have only one.